### DISCUSSION TITLE
ZHA automatic discovery of Zigbee coordinator bridges/gateways (Ethernet/WiFi network devices) that support Zeroconf or SSDP?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#667] [ZHA development]
Will you please consider adding Zeroconf discover or SSDP automatic network scanning and discovery to Home Assistant's ZHA integration?

Home Assistant OS/Supervisor already have some discovery features and functions needed for automatically discovering some types of network services and devices, and each integration for Home Assistant can choose to optionally opt-in to utilize these automatic network discovery features/functions by adding Zeroconf section and SSDP section, etc. to their manifest.json as described in their documentation:

https://www.home-assistant.io/integrations/zeroconf/

https://www.home-assistant.io/integrations/ssdp/

Reason for this "Service Discovery" / "Network Discover" request to Home Assistant's ZHA developers that seeing more and more [ser2net](https://linux.die.net/man/8/ser2net) or [com0com](http://sourceforge.net/projects/com0com/)/[com2tcp](http://astrogeeks.com/AstroGeeks/COM2TCP/index.html) type network-attached Zigbee adapters (a.k.a. Zigbee bridge/gateway) that runs a serial-to-IP proxy-server application that basically only allows it to tunnel a serial connection over your local home network (a.k.a. ser2net type serial to TCP/IP bridges/proxy appliance server devices) to allow you to place the Zigbee coordinator for ZHA in the most optimal location.

From an end-user experience point-of-view, it would be very user-friendly if such devices could be automagically discovered by Home Assistant during the ZHA integration's config flow for the initial installation and the adding of a network-attached Zigbee adapter.

As all probably know, ZHA's UI config flow already allow to manually config serial device path to use a TCP/IP network socket (using socat): 

- ESPHome based bridge: `socket://[IP]:[PORT]` or `socket://[FQDN]:[PORT]` (if got local DNS), example `socket://tube_zb_gw.local:6638`
- Sonoff ZBBridge if its ESP8266 hacked with Z2T (Zigbee2Tasmota): `socket://[IP]:[PORT]` for example `socket://192.168.1.11:8888`
- ZiGate Gateway: `socket://[IP]:[PORT]` for example `socket://192.168.1.10:9999`
- Xiaomi DGNWG05LM and ZHWG11LM gateways with Openlumi OpenWrt firmware: `socket://[IP]:[PORT]` for example `socket://192.168.1.10:9999`
- Tuya TYGWZ-01 and Lidl Silvercrest gateways hacked for serial-to-ip: `socket://[IP]:[PORT]` for example `socket://192.168.1.11:8888`

I would think that the discovery features that is missing there and needs to be added are automatically discovered by the ZHA integration is the IP + TCP-port + radio-type + baud rate port speed + data flow control method, as only then the normal zigpy radio library probe methods could step in and take it from there to do its probing for the Zigbee radio firmware?

Anyway, such devices addressed via IP are already supported without auto-discovery by ZHA and zigpy and today include:

- **Tube's Zigbee Gateways** and **Zigbee to Ethernet DIY Coordinator** by @tube0013 based on [ESPHome firmware](https://esphome.io/) with Zeroconf discover.
  - https://github.com/tube0013/tube_gateways
    - https://www.tubeszb.com/shop/coordinators/2
    - https://community.home-assistant.io/t/zigbee-3-0-to-ethernet-bridge-gateway-wired-not-wifi-inexpensive-diy-project-for-zha/257020/5
    - https://github.com/zigpy/zigpy/discussions/584
- **ZiGate WiFi Gateway** which firmware is based on [Jeelabs ESP-LINK firmware](https://github.com/jeelabs/esp-link/) serial bridge feature that supports Zeroconf discover.
  - https://zigate.fr/produit/zigate-pack-wifi-v1-3/
- **[Zigbee2Tasmota (Z2T)](https://tasmota.github.io/docs/Zigbee/)** on [Sonoff ZBBridge](https://www.digiblur.com/2020/07/how-to-use-sonoff-zigbee-bridge-with.html) or DIY, which firmware is based on Tasmota that already support SSDP discover.
  - https://tasmota.github.io/docs/Zigbee/
- Tuya TYGWZ-01 / Lidl Silvercrest Smart Gateway (Zigbee bridge with wired Ethernet port) hacked with serial-to-ip server.
  - https://paulbanks.org/projects/lidl-zigbee/ha.html
  - https://paulbanks.org/projects/lidl-zigbee/
  - https://github.com/banksy-git/lidl-gateway-freedom
  - https://github.com/zigpy/zigpy/discussions/650
  - https://community.home-assistant.io/t/hacking-the-silvercrest-lidl-tuya-smart-home-gateway/270934
- **OpenLumi** (OpenWrt firmware on hacked DGNWG05LM and ZHWG11LM gateways with ZiGate Zigbee stack).
  - https://openlumi.github.io/
  - https://github.com/openlumi
- Possibly a new ESP32 + EFR32MG21 based Zigbee to Ethernet Coordinator being working on by grobasoz
  - https://github.com/grobasoz
- DIY bridge using ESP32, ESP8266, or Arduino with Ethernet and firmware with a serial port to TCP converter code, like ex. [ESPHome](https://esphome.io/).
  - https://www.hackster.io/techbase_group/arduino-esp32-serial-port-to-tcp-converter-via-wifi-66d341
- DYI bridge using remote computer (like a Raspberry Pi Zero) running either a [ser2net](https://linux.die.net/man/8/ser2net) or a [com0com](http://sourceforge.net/projects/com0com/)/[com2tcp](http://astrogeeks.com/AstroGeeks/COM2TCP/index.html) configuration.
  - https://www.zigbee2mqtt.io/how_tos/how_to_connect_to_a_remote_adapter.html

If support for  Zeroconf and SSDP could be added to ZHA inside HA first then it might be more incentive to add the discovering of those?

ESPHome is popular in the HA community and used by Tube's Zigbee Gateways who sells Silabs + Texas Instruments based Zigbee gateways.

https://www.home-assistant.io/integrations/esphome/ 

"_ESPHome can be auto-discovered by Home Assistant. If an instance was found, it will be shown in the top of the list of integrations as “Discovered”. If that is the case click on the Configure button to start setting up the discovered instance._"

Tasmota based gateways could similarly be automatically discovered if the Tasmota device is configured for native discovery by the user:

https://www.home-assistant.io/integrations/tasmota/

ZiGate Gateway firmware does feature Zeroconf and the old ZiGate integration by @doudz also had for Zeroconf discover as a feature:

https://github.com/doudz/zigate/blob/master/zigate/transport.py#L385

zigpy/zigpy-zigate#36

**Comment by Hedda** [comment]
mDNS and/or UPnP could be alternative network discovery methods but probably not as suited for Zigbee ser2net type bridges, or?

https://www.home-assistant.io/integrations/discovery/

Zeroconf and/or SSDP is presumable better suited and again ZiGate WiFi and Tube's Zigbee Gateways already supports those, or?

https://developers.home-assistant.io/docs/creating_integration_manifest/#zeroconf

https://developers.home-assistant.io/docs/creating_integration_manifest/#ssdp

@tube0013 As you make Tube's Zigbee Gateways, what's your opinion on ZHA auto-discovering of ZB network gateways/bridges?

If makers of Ethernet bridges/gateways like yourself ship your pre-flashed with DHCP enabled they should become plug-and-play.

**Comment by Hedda** [comment]
Home Assistant founders have in related news now apparently now taken over leadership of ESPHome firmware core development:

https://www.home-assistant.io/blog/2021/03/18/nabu-casa-has-acquired-esphome/

They said on the Home Assistant Podcast that the project will remain open source 

https://hasspodcast.io/ha083/

Paulus hinted there that their focus will be on making it easier and more user-friendly to get up and running with ESPHome FW.

**Reply by MattWestb** [reply]
Wot have my neighbor Otto done !!

I was not surprising then it have growing to big for one man to manager and its have being seen the last year but hi have doing one great work getting ESPHome as far with limited resources.

I like it then its one easy way getting advances function connected to HA and its autonomy function that is making it working if HA is having one bad day (or night).

Perhaps getting one ser2net that is preconfigured to connecting to ZHA and only need to clicking download to your ESP and perhaps doing some encryption on the communication between the ESP and HA for getting it secure from unwanted guests :-)) 

So perhaps @tube013 is getting one fast lane integration for his Ethernet coordinators.

**Reply by Hedda** [reply]
> Perhaps getting one ser2net that is preconfigured to connecting to ZHA and only need to clicking download to your ESP

As mentioned above, ESPHome could be auto-discovered by the ZHA integration Home Assistant if integrations developers choose to opt-in Zeroconf and SSDP, but still, would also need to add functions to inform ZHA that the device can be used as a compatible Zigbee coordinator and what config it should be using; like: socket port, radio type,  baud rate port speed, and data flow control, etc.

https://www.home-assistant.io/integrations/esphome/ 

"_ESPHome can be auto-discovered by Home Assistant. If an instance was found, it will be shown in the top of the list of integrations as “Discovered”. If that is the case click on the Configure button to start setting up the discovered instance._"

> perhaps doing some encryption on the communication between the ESP and HA for getting it secure from unwanted guests 

I think questions about encrypted communication between gateway and ZHA is a different discussion for a other thread.

**Comment by Adminiuga** [comment]
Send PRs

**Comment by Hedda** [comment]
FYI; dmulcahey and tube0013 have now added Zeroconf discovery for Tube's Zigbee coordinators to ZHA in Home Assistant:

https://github.com/home-assistant/core/pull/48420

Nice!

**Comment by Hedda** [comment]
Do you know if Zeroconf in Home Assistant support DNS-SD (Rendezvous) "TXT record" to pass config parameters to integrations?

[http://www.zeroconf.org/Rendezvous/txtrecords.html](http://www.zeroconf.org/Rendezvous/txtrecords.html)

If so, you think that Zeroconf DNS-SD (DNS Service Discovery) could pass "TXT records" data on to the ZHA integration config flow?

Example of config parameters that Zeroconf could be passed along to ZHA integration via TXT records:

* radio_type=ezsp
* baud_rate_speed=115200
* data_flow_control=software
* tcp_port_serial_gateway=8888

I understand initial PR https://github.com/home-assistant/core/pull/48420 for ZHA rely on explicit host name for discovery but maybe TXT record is better?

TXT records should be supported by mDNS and Zeroconf standards if I read all this correctly:

[https://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-based_service_discovery](https://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-based_service_discovery)

 "_Each service instance is described using a DNS SRV and **DNS TXT record**. A client discovers the list of available instances for a given service type by querying the DNS PTR record of that service type's name; the server returns zero or more names of the form <Service>.<Domain>, each corresponding to a SRV/TXT record pair. The SRV record resolves to the domain name providing the instance, while **the TXT can contain service-specific configuration parameters.**_"

[https://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-SD_with_multicast](https://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-SD_with_multicast)

"_DNS Service Discovery (DNS-SD) requests can also be sent using mDNS to yield zero-configuration DNS-SD. This uses DNS PTR, SRV, **TXT records** to advertise instances of service types, domain names for those instances, and optional configuration parameters for connecting to those instances. But SRV records can now resolve to .local domain names, which mDNS can resolve to local IP addresses._"

[https://en.wikipedia.org/wiki/TXT_record](https://en.wikipedia.org/wiki/TXT_record)

"_**TXT record** (short for text record) is a type of resource record in the Domain name system (DNS) used to provide the ability to associate arbitrary text with a host or other name, such as human readable information about a server, network, data center, or other accounting information. It is also often used in a more structured fashion to record small amounts of machine-readable data into the DNS_."
